### ISSUE TITLE
FATAL:   While performing build: packer failed to pack: while unpacking tmpfs: error unpacking rootfs: mkdir rootfs

### ISSUE THREAD

**Issue by vbeliveau** [apptainer/apptainer#1800]
### Version of Apptainer

What version of Apptainer (or Singularity) are you using? Run

`apptainer version 1.2.4-1.el7`

### Expected behavior

SIF image is build correctly.

### Actual behavior

What actually happened? Why was it incorrect?

```
> apptainer build --fakeroot test.sif test.def
INFO:    Starting build...
Getting image source signatures
Copying blob 578acb154839 done  
Copying config fd0ea473a1 done  
Writing manifest to image destination
Storing signatures
FATAL:   While performing build: packer failed to pack: while unpacking tmpfs: error unpacking rootfs: mkdir rootfs: mkdir /staff/vbeliveau/apptainer-tmp/build-temp-1437880609/rootfs: no such file or directory
```
Dummy content of `test.def`:
```
Bootstrap: docker
From: debian:bookworm-slim

%post
    # Some command requiring root
    apt-get update -q
```

### What OS/distro are you running

```
NAME="openSUSE Leap"
VERSION="15.5"
ID="opensuse-leap"
ID_LIKE="suse opensuse"
VERSION_ID="15.5"
PRETTY_NAME="openSUSE Leap 15.5"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:leap:15.5"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
DOCUMENTATION_URL="https://en.opensuse.org/Portal:Leap"
LOGO="distributor-logo-Leap"
```

### How did you install Apptainer
Using [install-unprivileged.sh](https://apptainer.org/docs/admin/main/installation.html#install-unprivileged-from-pre-built-binaries).

### Note
This problem came up as part of solving issue #1751.

**Comment by DrDaveD** [comment]
Unfortunately I have not been able to reproduce this.

I started an opensuse/leap docker container on an el7 host including bind-mounting my home directory, created an unprivileged account inside it matching my own on the host, used install-unprivileged.sh to install apptainer, and did the test build.  It all worked fine for me.

Can you think of something else about your environment that might be contributing to this?  My home directory is on NFS, is yours on a special filesystem?  Does it make a difference if you use `APPTAINER_TMPDIR` or `APPTAINER_CACHEDIR` to move your temporary or cache directory to other places, especially other filesystem types?  For that matter, are you already setting `APPTAINER_TMPDIR=/staff/vbeliveau/apptainer-tmp`?  If so, what happens if you don't?

**Comment by vbeliveau** [comment]
@DrDaveD I've tried to build the test image again on a different server (with the same configuration) and this time the build succeeded without problem. The main difference between the two systems is that a fakeroot mapping to my user had been configure (using `apptainer config fakeroot --add <user>`) on the server where the build fails and not on the other where it succeeds. I wasn't expecting fakeroot to work at all without first setting up the fakeroot mapping, but it seems that it does anyway when using the unprivileged install?

Here is the fakeroot mapping on the server where it is actually set up (same as in #1751):
```
> cat /etc/subuid
4290:4294836224:65536
> cat /etc/subgid
4290:4294836224:65536
```

**Comment by DrDaveD** [comment]
I suspect that's not the real difference, because it succeeded for me on a system with /etc/subuid mapping.  You can have apptainer ignore the subuid mapping with an (undocumented) run/shell/exec `--ignore-subuid` option to find out.

**Comment by vbeliveau** [comment]
@DrDaveD the build does succeed on the server where it was failing when using `--ignore-subuid`. I guess this does narrow it down to an issue with the  /etc/subuid mapping?

When I run without `--ignore-subuid`, I get the following debug output:
```
DEBUG   [U=0,P=1]          unpackRootfs()                setting umoci rootless mode
DEBUG   [U=0,P=1]          cleanUp()                     Cleaning up "/staff/vbeliveau/apptainer-tmp/build-temp-1460096801/rootfs" and "/staff/vbeliveau/apptainer-tmp/bundle-temp-1971055350"
FATAL   [U=0,P=1]          runBuildLocal()               While performing build: packer failed to pack: while unpacking tmpfs: error unpacking rootfs: mkdir rootfs: mkdir /staff/vbeliveau/apptainer-tmp/build-temp-1460096801/rootfs: no such file or directory
DEBUG   [U=0,P=3061]       func1()                       Wait for process 3084 complete with status 65280, error <nil>
DEBUG   [U=0,P=3061]       Master()                      Child exited with exit status 255
```

However, when running with `--ignore-subuid` I have:
```
DEBUG   [U=0,P=4129]       unpackRootfs()                setting umoci rootless mode
```

In the case where it fails, the process running `unpackRootfs()` is `P=1` (the init process?), whereas in the other case it's `P=4129` (what I assume to be a newly created process).

**Comment by DrDaveD** [comment]
Yes, I see that too without `--ignore-subuid` and yet it doesn't fail for me when I run opensuse/leap under docker.  It must be running with the equivalent of an action command's `--pid` option to give it a separate process namespace.  Maybe it's a difference in the kernel, or maybe there's something else different about the conditions on that machine.

**Comment by exaos** [comment]
Same issue here. My environment:

- macOS 15.1.1, M2
- OrbStack 1.8.2
- NixOS 24.05
- apptainer 1.3.3 (installed with `nix-env -iA nixos.apptainer`)

Procedure to repeat the issue:
- run `orb` and jump into nixos
- just run `apptainer build alpine.sif docker://alpine`

I got the following error message:

```
[vivodo@nixos:/Volumes/Dongting/swamp/build/apptainer]$ apptainer build --ignore-subuid alpine.sif docker://alpine
INFO:    Starting build...
Copying blob 9986a736f7d3 skipped: already exists
Copying config 511a44083d done   |
Writing manifest to image destination
2024/11/23 16:08:05  info unpack layer: sha256:9986a736f7d3d24bb01b0a560fa0f19c4b57e56c646e1f998941529d28710e6b
2024/11/23 16:08:05  info verified reader: 3076014 bytes of trailing data discarded from vrdr[/Volumes/Dongting/swamp/build/apptainer/tmp/bundle-temp-832640174/blobs/sha256/9986a736f7d3d24bb01b0a560fa0f19c4b57e56c646e1f998941529d28710e6b]
FATAL:   While performing build: packer failed to pack: while unpacking tmpfs: error unpacking rootfs: unpack layer: unpack entry: bin/arch: apply hdr metadata: clear xattr metadata: /Volumes/Dongting/swamp/build/apptainer/tmp/build-temp-3168490444/rootfs/bin/arch: unpriv.lclearxattrs: unpriv.llistxattr: no such file or directory

```

**Comment by DrDaveD** [comment]
That's  a different error message and a much different environment.  I suggest making a separate issue, although it's going to be very difficult for us to reproduce because the environment is so unusual.
